### PR TITLE
Long session timeout issue

### PR DESCRIPTION
**Pull request by cammckenzie**
Fix for issue where LOST events were created at 4/3 of the session timeout rather than at the end of session timeout.

### CODE REVIEW

File: curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java [hunk 1]
```diff
@@ -252,7 +252,9 @@ private void processEvents()
             {
                 int lastNegotiatedSessionTimeoutMs = client.getZookeeperClient().getLastNegotiatedSessionTimeoutMs();
                 int useSessionTimeoutMs = (lastNegotiatedSessionTimeoutMs > 0) ? lastNegotiatedSessionTimeoutMs : sessionTimeoutMs;
-                int pollMaxMs = (useSessionTimeoutMs * 2) / 3; // 2/3 of session timeout
+                long elapsedMs = startOfSuspendedEpoch == 0 ? 0 : System.currentTimeMillis() - startOfSuspendedEpoch;
+                long pollMaxMs = useSessionTimeoutMs - elapsedMs;
```
Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java]: We should probably add a 50% factor here or something. It seems suspicious to wait the entire session time if there isn't a suspended state.

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessSemaphore.java [hunk 2]
```diff
@@ -148,10 +148,9 @@ else if ( newState == ConnectionState.RECONNECTED  )
                     }
                 }
             });
-
-            timing.sleepABit();
+      
             server.stop();
-            Assert.assertTrue(timing.awaitLatch(lostLatch));
+            Assert.assertTrue(timing.multiple(2).awaitLatch(lostLatch));
```
Comment (Randgalt) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/locks/TestInterProcessSemaphore.java]: Maybe timing.multiple(1.25) or something to see how close we can get.

File: curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java (hunk 1 above)
Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java]: What do you mean if there isn't a suspended state? Shouldn't we go into a suspended state as soon as the connection is dropped? Then, we will wait up to session timeout for an event of any type. If we don't get one, then we assume the session is lost and inject a LOST event.

Comment (Randgalt) [inline curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java]: If startOfSuspendedEpoch is 0, the wait is useSessionTimeoutMs. I'm thinking that if startOfSuspendedEpoch the max wait should (useSessionTimeoutMs / 2) or something. It seems safer to me.

Comment (cammckenzie) [inline curator-framework/src/main/java/org/apache/curator/framework/state/ConnectionStateManager.java]: Ok, sounds reasonable.

**Comment by cammckenzie** [comment]
Ok, I've pushed some additional changes to fix your comments. Are you happy for me to merge?

**Comment by Randgalt** [comment]
LGTM

**Pull request merged**